Feature/rob 378 sonar rov integration

hugoyvrn opened the pull request:
Documentation on https://github.com/forssea-robotics/frs_docs/blob/develop/sensors/sonar/index.rst is ongoing.
This PR contains: 
- Better handling of ros parameters (mainly by taking into account oculus sonar feedbacks).
- A better handling of standby mode (taking into account overheating and a ros param).
- @JaouadROS code for conic image displaying.
- A better code architecture and dependency update.
- Fixed bugs
- And also remaining bugs and no finished feature (publish bearings on a topic, optimization on standby mode handling, etc).


@JaouadROS @AugusteBourgois

Commit:
deguging import rosbag to ros2bag

Commit:
Start script bag_to_oculus

Commit:
Update dev scripts and test for oculus_to_bag

Commit:
Simple rosbag creator with custom timestamp

Commit:
Rosbag creator for OculusStampedPing message

Commit:
generate a rosbag ok. Start of reading .oculus in a c++ parser

Commit:
.oculus parser: get a item header: seams good

hugoyvrn committed:
parser ok

hugoyvrn committed:
Bug correction

hugoyvrn committed:
Adding chatGPT advices for timestamp

hugoyvrn committed:
oculus_to_bag update : need to handle time stamp

hugoyvrn committed:
oculus to bag ok

hugoyvrn committed:
Minor update and bug correction

hugoyvrn committed:
Create scripts for display oculus data

hugoyvrn committed:
Updating oculus ros message field in oculus_to_rosbag

hugoyvrn committed:
Changing oculus message type from OculusPing to Ping

hugoyvrn committed:
Image publisher : working badly (need gain correction probably

hugoyvrn committed:
oculus_subscriber_to_image okay

hugoyvrn committed:
oculus_subscriber_to_image correction

hugoyvrn committed:
tmp install tuto commit

Commit:
Inverse gray scale

langlath committed:
possibilité de choisir la fréquence du publisher :snail:

langlath committed:
merging :shipit:

Commit:
Debugage oculus_sonar_node with oculus_driver update

Commit:
merge

langlath committed:
merging

langlath committed:
Merge branch 'rosbag_dev' of https://github.com/hugoyvrn/oculus_ros2 into rosbag_dev

awanctin committed:
ajout des gains dans les Ping et début de débuggage de count suscribers

Commit:
Debugging oculus_ros2 to exiting standby mode done

Commit:
oculus_subscriber_to_image.py not working with gain

langlath committed:
Merge branch 'rosbag_dev' of https://github.com/hugoyvrn/oculus_ros2 into rosbag_dev

awanctin committed:
Gains ok, image malformée

langlath committed:
working but not correct version of oculus_subscriber_to_image

langlath committed:
working but not correct version of oculus_subscriber_to_image

langlath committed:
clean version without print

langlath committed:
adding parameter freq, working

awanctin committed:
image avec gains affichée mais gains pas pris en compte

awanctin committed:
comparaison 2 images

Commit:
Imgae size texts if no gain

langlath committed:
commit pour antoine

langlath committed:
Merge branch 'rosbag_dev' of https://github.com/hugoyvrn/oculus_ros2 into HEAD

Commit:
working ugly modification seems working in real time

langlath committed:
cleaner version of subscriber_to_image.py

awanctin committed:
réparation pour avoir la traj et les nuages de points sur rviz

awanctin committed:
oculus marche en temps réel mais pas avec octomap

Commit:
Minor cleaning

hugoyvrn committed:
Parameters seems to work. Need tests and code clean up

hugoyvrn committed:
Adding a test plan

hugoyvrn committed:
Some cleaning and image publisher corrections

hugoyvrn committed:
Updating test plan (not finish)

hugoyvrn committed:
Some cleaning and bug corrections on ros param update

hugoyvrn committed:
Update tests plan

hugoyvrn committed:
Minor bug corrections

hugoyvrn committed:
General update. WIP

hugoyvrn committed:
ROS parameter refactoring and other update

hugoyvrn committed:
Bugs correction

hugoyvrn committed:
Keep ros parameters in memory and update ros parameters to keep it in coherence of the sonar configuration. Not tested

hugoyvrn committed:
Adding a parameter to handle standby mode

hugoyvrn committed:
Ready for test.

hugoyvrn committed:
Driver only for M1200d oculus

hugoyvrn committed:
Standby mode bug correction

hugoyvrn committed:
Importing ros1 viewer node from https://github.com/JaouadROS/Sonar_processing_display/blob/image_processing_working_version/src/oculus_viewer_node.cpp

hugoyvrn committed:
WIP Addapting jaouad fan visualisation

hugoyvrn committed:
Ready for tests in bucket

hugoyvrn committed:
Bug correction

hugoyvrn committed:
SonarViewer constructor

hugoyvrn committed:
Standby mode initialisation issue

hugoyvrn committed:
Adding pressure and temprature publisher

hugoyvrn committed:
Setting static parameters

hugoyvrn committed:
Some cleaning

hugoyvrn committed:
Using tempalate function in order to build with c++17

hugoyvrn committed:
Wrong fan image

hugoyvrn committed:
restructure image viewer in a separate node (optional)

hugoyvrn committed:
Adding pre-commit

hugoyvrn committed:
Some cleaning

JaouadROS committed:
minimal tweaking code to get it work

JaouadROS committed:
cleaning code

JaouadROS committed:
quick optimization

hugoyvrn committed:
WIP code crash when initializing range parameter

hugoyvrn committed:
White on back ground of conic view

hugoyvrn committed:
Code cleaning

hugoyvrn committed:
Adding corpyrights

hugoyvrn committed:
WIP creating metapackage

hugoyvrn committed:
WIP creating metapackage

hugoyvrn committed:
Handleling oculus_driver dependancy

hugoyvrn committed:
Handeling oculus_driver dependancy

hugoyvrn committed:
Handeling oculus_driver dependancy

hugoyvrn committed:
Handeling oculus_driver dependancy

hugoyvrn committed:
Handeling oculus_driver dependancy

hugoyvrn committed:
Some cleaning

hugoyvrn committed:
Merge pull request #1 from hugoyvrn/feature/ROB-378

Feature/rob-378

hugoyvrn committed:
Some cleaning

hugoyvrn committed:
Some bug correction

hugoyvrn committed:
Tmp fix of bug

hugoyvrn committed:
Merge

hugoyvrn committed:
Some renaming

hugoyvrn committed:
WIP handleling 16 bits data depth in image viewer

hugoyvrn committed:
Code cleaning

AugusteBourgois commented on README.md:
```diff
@@ -7,14 +7,13 @@ ROS2 node for the Blueprint Subsea Oculus sonar.
-* CMake 3.23
+* Ubuntu 22.04 LTS
+* ROS2 humble
+* CMake ???
```
3.18 min

AugusteBourgois commented on oculus_interfaces/package.xml:
```diff
@@ -2,7 +2,7 @@
     <name>oculus_interfaces</name>
     <version>0.0.1</version>
-    <description>TODO: Package description</description>
+    <description>TODO(hugoyvrn): Package description</description>
```
Fill in the description

AugusteBourgois commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "oculus_driver/oculus_driver"]
```
We shouldn't use git submodules. The correct way of handling external dependencies is to add them as dependencies and let the CI build them. Locally, you should just clone the repository in your workspace and build with colcon directly.

AugusteBourgois replied:
So you can remove the folder oculus_driver and its content

hugoyvrn replied:
Hmm. I have two questions:
- Does this repo link to the CI ? (I tought not)
- Then how I set up the build as I want ? In the CI ?

AugusteBourgois commented on README.md:
```diff
@@ -42,54 +41,86 @@ cd .. && colcon build
-Clone or copy the oculus_driver library :
+This library follows a standard CMake compilation procedure. To use oculus_driver you have two possibility :
+
+#### Let colcon handle the dependency
```
This should not be an option. External dependencies are handled by the CI.

AugusteBourgois commented on README.md:
````diff
@@ -42,54 +41,86 @@ cd .. && colcon build
-```
-echo $CMAKE_PREFIX_PATH
+
+#### install the python package
````
We will never use the python bindings, so no need to document them. Instead, if possible document how not to build them.

AugusteBourgois replied:
Ok, my bad, I did not notice that.

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+find_package(cv_bridge REQUIRED)
+find_package(OpenCV REQUIRED)
+
+# find_package(oculus_driver QUIET)
```
Remove unnecessary comments

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+find_package(oculus_driver REQUIRED)
+
 
 add_executable(oculus_sonar_node
```
You should regroup your commands by executable.

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+# target_include_directories(oculus_sonar_node PRIVATE
+#   ${CMAKE_CURRENT_SOURCE_DIR}/include
+#   ${OpenCV_INCLUDE_DIRS})
+target_include_directories(oculus_sonar_node PUBLIC
```
```cpp
target_include_directories(oculus_sonar_node PUBLIC
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
    $<INSTALL_INTERFACE:include>)
```

AugusteBourgois replied:
Same for oculus_viewer_node

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+)
+
+include_directories( ${OpenCV_INCLUDE_DIRS} )
```
Remove

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+include_directories( ${OpenCV_INCLUDE_DIRS} )
+
+target_link_libraries(oculus_sonar_node PRIVATE
+  ${OpenCV_LIBRARIES}
```
Use `${OpenCV_LIBS}`

AugusteBourgois replied:
I don't think that you need `${ament_LIBRARIES}`

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+  ${ament_LIBRARIES}
     oculus_driver
 )
 target_compile_features(oculus_sonar_node PRIVATE cxx_std_17)
```
Remove, and add 
```cmake
if(NOT CMAKE_CXX_STANDARD)
  set(CMAKE_CXX_STANDARD 17)
endif()

if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
  add_compile_options(-Wall -Wextra -Wpedantic -ggdb3)
endif()
```

juste below the `project` line of the CmakeLists.txt

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+    oculus_interfaces
+    rcl_interfaces
+    sensor_msgs
+    OpenCV
```
You should either use `ament_target_dependencies(... PUBLIC OpenCV)` or target_link_library + target_include_directory. They are otherwise redundant

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
-ament_package()
+install(PROGRAMS
+    scripts/display_oculus_file.py
+    RENAME display_oculus_file
```
No need to rename

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+)
+install(PROGRAMS
+    scripts/oculus_to_rosbag.py
+    RENAME oculus_to_rosbag
```
No need to rename

AugusteBourgois commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
+    oculus_viewer_node
+    DESTINATION lib/${PROJECT_NAME}
+)
+
```
Make you install commands one-liners

AugusteBourgois commented on oculus_ros2/cfg/default.yaml:
```diff
@@ -1,48 +1,41 @@
     # 2: High frequency (2.1MHz, narrow aperture).
 
-    ping_rate: 0 # Frequency of ping fires.
+    ping_rate: 0 # Frequency of ping fires. Default value is 2.
```
Then set it to to in this file maybe ?

AugusteBourgois commented on oculus_ros2/cfg/tmp.yaml:
```diff
@@ -0,0 +1,41 @@
+oculus_sonar:
```
Why is there a tmp file ?

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/conversions.hpp:
```diff
@@ -0,0 +1,140 @@
+
+namespace oculus {
+
+inline rclcpp::Time to_ros_stamp(const SonarDriver::TimePoint& stamp) {
```
You can rename all these functions to `toMsg` to match the usual ros API

hugoyvrn replied:
I rename all `to_ros_stamp` and all `copy_to_ros` in `toMsg` ?

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/oculus_sonar_node.hpp:
```diff
@@ -0,0 +1,142 @@
+#include <sensor_msgs/msg/fluid_pressure.hpp>
+#include <sensor_msgs/msg/temperature.hpp>
+
+typedef struct {
```
```cpp
struct RosParameters {
  int frequency_mode;
  int ping_rate;
  int data_depth;
  int nbeams;
  bool gain_assist;
  double range;
  int gamma_correction;
  double gain_percent;
  double sound_speed;
  bool use_salinity;
  double salinity;
};
```

AugusteBourgois replied:
Rename to `SonarParameters`

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/oculus_sonar_node.hpp:
```diff
@@ -0,0 +1,142 @@
+  rosParameters currentRosParameters;
+  oculus::SonarDriver::PingConfig currentConfig;
+
+  bool is_in_run_mode;  // Same value as ros parameter "run"
```
Rename to `is_running`

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/oculus_sonar_node.hpp:
```diff
@@ -0,0 +1,142 @@
+  rcl_interfaces::msg::SetParametersResult set_config_callback(const std::vector<rclcpp::Parameter>& parameters);
+
+  void enable_run_mode();
+  void desable_run_mode();
```
Typo `disable_run_mode`

hugoyvrn replied:
Moy bi

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/oculus_sonar_node.hpp:
```diff
@@ -0,0 +1,142 @@
+}
+
+template <class T>
+void OculusSonarNode::update_ros_config_for_param(
```
Why do you have ros_param_name and param_name ? Can't they be the same ?

hugoyvrn replied:
Usually yes. But not for all parameters :
 - the frequency is named `master_mode` is oculus_driver API and `frequency_mode` in ros parameter;
 - the sound of speed is named `speed_of_sound_used` is oculus_driver API and `sound_speed` in ros parameter.

use at https://github.com/forssea-robotics/oculus_sonar/blob/feature/ROB-378-sonar-rov-integration/oculus_ros2/src/oculus_sonar_node.cpp#L504 

Do you want me to rename those two parameters?

AugusteBourgois replied:
You don't need to display the name of the parameter used in the API. You can just use your own name that makes more sense as a ros parameter, and change the appropriate one in the API.

hugoyvrn replied:
It is already like that, I made a mistake. Thanks

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/oculus_sonar_node.hpp:
```diff
@@ -0,0 +1,142 @@
+    const T& old_val,
+    const T& new_val,
+    const std::string& param_name,
+    const std::string& param_name_to_display) const {
```
Why having this extra parameter ?

hugoyvrn replied:
Idem than for update_ros_config_for_param

    the frequency is named master_mode is oculus_driver API and frequency_mode in ros parameter;
    the sound of speed is named speed_of_sound_used is oculus_driver API and sound_speed in ros parameter.

use at https://github.com/forssea-robotics/oculus_sonar/blob/feature/ROB-378-sonar-rov-integration/oculus_ros2/src/oculus_sonar_node.cpp#L504

AugusteBourgois replied:
You don't need to display the name of the parameter used in the API. You can just use your own name that makes more sense as a ros parameter, and change the appropriate one in the API.

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_publisher_;
+};
+
+template <typename T>
```
This is better : [https://stackoverflow.com/questions/55102504/populate-a-vector-with-linearly-increased-values](https://stackoverflow.com/questions/55102504/populate-a-vector-with-linearly-increased-values)

hugoyvrn replied:
Like that :
```C++
template <typename T>
std::vector<double> linspace(T start_in, T end_in, int num_in) {
  const auto start = static_cast<double>(start_in);
  const auto end = static_cast<double>(end_in);
  const auto num = static_cast<double>(num_in);
  std::vector<double> linspaced(std::max(0,num_in-1));

  if (num == 0) {
    return linspaced;
  }
  if (num == 1) {
    linspaced.push_back(start);
    return linspaced;
  }

  double delta = (end - start) / (num - 1);
  std::generate(linspaced.begin(), linspaced.end(), [i = 0, &delta]() mutable { return i++ * delta; });
  linspaced.push_back(end);  // I want to ensure that start and end
                             // are exactly the same as the input
  return linspaced;
}
```
?

AugusteBourgois replied:
If this method is only used once in your code, you should aim at removing it and using the line `std::generate(linspaced.begin(), linspaced.end(), [i = 0, &delta]() mutable { return i++ * delta; });` directly. No need to make it extra safe by adding checks etc...
No need to use templates either.

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+  std::vector<double> ranges = linspace(0., ping_range, height);
+  int image_width = 2 * std::sin(bearing * M_PI / 180) * ranges.size();
+  cv::Mat mono_img;
+  if (std::is_same<dataType, uint8_t>::value) {
```
```cpp
if constexpr (std::is_same<dataType, uint8_t>::value)
```

is probably better, as this will be chosen at compile time and not at runtime

AugusteBourgois replied:
Use `std::numeric_limits<uint8/16_t>::max();` to avoid magic numbers

hugoyvrn replied:
I hadn't see your comment, so I used `1 << sizeof(DataType) * CHAR_BIT`. I think std::numeric_limits is doing the same.

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+  cv::Mat rawDataMat(height, (width + 4), CV_8U);
+  std::memcpy(rawDataMat.data, ping_data.data() + offset, height * (width + 4));
+
+  int bearing = (master_mode == 1) ? 65 : 40;
```
Magic numbers : create const variables with explicit names and use them.

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+  //   }
+  // }
+
+  const float ThetaShift = 1.5 * 180;
```
1.5 is a magic number

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+    mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_16UC1) * 65535;
+  }
+
+  // for (int i = 0; i < image_width; i++) {
```
Remove if unnecessary

AugusteBourgois commented on oculus_ros2/launch/default.launch.py:
```diff
@@ -8,32 +14,54 @@
-        description='Filters Configuration'))
-    
+
+    # ld.add_action(
```
Remove if unnecessary

AugusteBourgois commented on oculus_ros2/launch/default.launch.py:
```diff
@@ -8,32 +14,54 @@
-      )
-   
+        get_package_share_directory("oculus_ros2"), "cfg", "default.yaml"
+        # get_package_share_directory("oculus_ros2"), "cfg", "tmp.yaml"
```
Remove if unnecessary

AugusteBourgois commented on oculus_ros2/package.xml:
```diff
@@ -13,21 +13,35 @@
   <build_depend>roscpp</build_depend>
   <build_depend>rospy</build_depend>
   <build_depend>std_msgs</build_depend>
+  <build_depend>sensor_msgs</build_depend>
```
All the build + build_export + exec dependencies can be listed as <depend>

hugoyvrn replied:
I think your sentence is not finished, in any case I do not understand what I can do.

AugusteBourgois replied:
All the build + build_export + exec dependencies can be listed as `<depend>`

AugusteBourgois commented on oculus_ros2/package.xml:
```diff
@@ -13,21 +13,35 @@
 
   <depend>Boost</depend>
   <depend>oculus_interfaces</depend>
+  <!-- <depend>rqt_reconfigure</depend> -->
```
Remove what's unnecessary

hugoyvrn replied:
Ongoing

AugusteBourgois commented on oculus_ros2/scripts/display_oculus_file.py:
```diff
@@ -0,0 +1,192 @@
+
+if __name__ == "__main__":
+
+    main()
```
Why putting all the main code in a main function ?

hugoyvrn replied:
To put
```python
   parser = argparse.ArgumentParser(
        prog="OculusFileReader",
        description="Example of how to read and display the content of a .oculus "
        + "file. This will display the first ping from a the file.",
    )
    parser.add_argument("filename", type=str, help="Path to a .oculus file to display")
    parser.add_argument(
        "-r", "--rate", type=float, default=1e-2, help=". Default to 1e-2"
    )

    args = parser.parse_args()
    ```
    outside of it.

AugusteBourgois commented on oculus_ros2/scripts/oculus_to_rosbag.py:
```diff
@@ -0,0 +1,265 @@
+        self.new_topic(name=self.args.topicname, type="oculus_interfaces/msg/Ping")
+
+    def create_ros_oculus_msg(self, ping_msg):
+        # print("timestamp",           ping_msg.timestamp())
```
Remove what's unnecessary

AugusteBourgois commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -0,0 +1,194 @@
+    }
+  }
+
+  cv::Mat img = data;
```
Remove all magic number occurences

JaouadROS replied:
If this function `stream_and_filter` will be called inside `publish_fan` then this code is redundent, it is the same as:
```
cv::Mat rawDataMat(height, (width + 4), CV_8U);
std::memcpy(rawDataMat.data, ping_data.data() + offset, height * (width + 4));
```
So the conversion from `ping_data` vector to a `cv::Mat` matrix is done only once.

AugusteBourgois commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -0,0 +1,194 @@
+//                                     const int &offset,
+//                                     const std::vector<uint8_t> &ping_data,
+//                                     cv::Mat &data)
+void SonarViewer::stream_and_filter(const oculus::PingMessage::ConstPtr& ping, cv::Mat& data) {
```
This function should be documented

JaouadROS replied:
`stream_and_filter` is not called in your code, right? so not tested yet?

AugusteBourgois commented on oculus_ros2/src/oculus_sonar_node.cpp:
```diff
@@ -1,370 +1,558 @@
+    //  sonar_driver_(std::make_shared<SonarDriver>(this->io_service_.io_service())),
+    sonar_viewer(static_cast<rclcpp::Node*>(this)),
+    frame_id(this->declare_parameter<std::string>("frame_id", "sonar")),
+    topics_prefix(this->declare_parameter<std::string>("topics_prefix", "")),
```
Unnecessary topics_prefix parameter.

AugusteBourgois commented on oculus_ros2/src/oculus_sonar_node.cpp:
```diff
@@ -1,370 +1,558 @@
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+
+  if (!this->has_parameter("run")) {
```
Why not put `run` in the list_initialization ?

AugusteBourgois commented on oculus_ros2/src/oculus_sonar_node.cpp:
```diff
@@ -1,370 +1,558 @@
+  }
+
+  // Get the current sonar config
+  update_parameters(currentSonarParameters, this->sonar_driver_->current_ping_config());
```
If the function `update_parameters` is used to load download the sonars parameters, it should be called `download_parameters`

hugoyvrn replied:
update_parameters is used to update the local variables currentSonarParameters and currentRosParameters. this->sonar_driver_->current_ping_config() downloads the sonar parameters, this function is only used during initialization.

AugusteBourgois replied:
Then rename it `updateLocalParameters`

AugusteBourgois commented on oculus_ros2/src/oculus_sonar_node.cpp:
```diff
@@ -1,370 +1,558 @@
-    size_t seconds = nano / 1000000000;
-    return rclcpp::Time(seconds, nano - 1000000000*seconds);
+void OculusSonarNode::publish_status(const OculusStatusMsg& status) const {
+  if (status.partNumber != 1042)
```
Magic number

AugusteBourgois commented on oculus_ros2/src/oculus_sonar_node.cpp:
```diff
@@ -1,370 +1,558 @@
-    msg.header.stamp    = to_ros_stamp(this->sonar_driver_->last_header_stamp());
-    msg.header.frame_id = "oculus_sonar";
-    this->ping_publisher_->publish(msg);
+inline rclcpp::Time to_ros_stamp(const SonarDriver::TimePoint& stamp) {
```
This function already exists in conversions.h !

AugusteBourgois commented on oculus_ros2/src/oculus_sonar_node.cpp:
```diff
@@ -1,370 +1,558 @@
+    } else if (std::find(dynamic_parameters_names.begin(), dynamic_parameters_names.end(), param.get_name()) !=
+               dynamic_parameters_names.end()) {
+      send_param_to_sonar(param, result);
+    } else {  // TODO
```
Remove the else, and return true by default

AugusteBourgois reviewed (changes requested):
A first row of comments; Please address them all, then re run precommit, then use clang format in vscode and address the problems raised.

AugusteBourgois reviewed (changes requested):
A first row of comments; Please address them all, then re run precommit, then use clang format in vscode and address the problems raised.

JaouadROS commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+}
+
+template <typename dataType>
+void SonarViewer::publish_fan(const int& width,
```
Move `publish_fan` definition to `sonar_view.cpp` along with other members.

AugusteBourgois replied:
This will be difficult as this is a templated function

JaouadROS commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -0,0 +1,194 @@
+  cv::normalize(result, result, 0, 255, cv::NORM_MINMAX);
+  file_name2 << "/home/jaouadros/catkin_ws/src/Sonar_processing_display/tests/results/after/image_" << image_i << ".jpg";
+  cv::imwrite(file_name2.str(), result);
+  image_i++;*/
```
To be removed.

JaouadROS commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -0,0 +1,194 @@
+
+int grid_presence_counter[36][25] = {0};
+int grid_absence_counter[36][25] = {0};
+int frames_counter = 0;
```
To be removed.

AugusteBourgois replied:
Why ? What is this ?

hugoyvrn replied:
I think this is only used to filter the image. But the image is never filtered at the moment, so it is not useful (unless we want to add a filter option in the package). Moreover, I don't think that a gamma correction is a relevant filter from the oculus sonar data since there is already a gamma filter in the sonar software.

JaouadROS replied:
`gammaCorrection` is to adjust the brightness and contrast of an image. `type2str` is to print the image type. The other variables were used for obstacle detection algorithm.
We don't need all of that.

JaouadROS commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -0,0 +1,194 @@
+  for (int i = 0; i < result.cols; i++) result_shifted_rows.col(i).copyTo(result.col((i + shift) % result.cols));
+  result.setTo(0, result < 0);
+  cv::normalize(result, result, 0, 1, cv::NORM_MINMAX);
+  cv::imshow("result", result);
+  cv::waitKey(1);
```
To be removed when the function is tested.

hugoyvrn replied:
SonarViewer::streamAndFilter is not called. Do we want to add a filter one the package, or can I remove the entire function ? @AugusteBourgois

AugusteBourgois replied:
I don't know if we should leave the filtering code inside this code. I think the filtering should occur somewhere else. Down the processing line.

AugusteBourgois replied:
So I think you can remove the whole function. But keep it somewhere, in a TODO file for example, so that we don't have to reimplement it later.

JaouadROS commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -0,0 +1,194 @@
+
+  cv::Mat img = data;
+  img = img / 255.0;
+  cv::imshow("img", img);
```
To be removed when the function is tested.

JaouadROS commented on oculus_interfaces/package.xml:
```diff
@@ -2,7 +2,7 @@
-    <description>TODO: Package description</description>
+    <description>TODO(hugoyvrn): Package description</description>
     <maintainer email="auguste@forssea-robotics.fr">Auguste BOURGOIS</maintainer>
     <license>BSD3</license>
```
This code is open sourced, correct?

AugusteBourgois replied:
Yes, I think so

hugoyvrn replied:
Is BSD3  license compatible with forssea copyright ? I think it is.

AugusteBourgois replied:
Ok for me, we'll change it later if necessary

JaouadROS commented on oculus_ros2/CMakeLists.txt:
```diff
@@ -7,44 +7,112 @@ find_package(oculus_interfaces REQUIRED)
-endif()
+find_package(image_transport REQUIRED)
+find_package(cv_bridge REQUIRED)
+find_package(OpenCV REQUIRED)
```
Better to specify a min version for `OpenCV`. This package is supposed to work with `Humble`, so you can have its OpenCV version as your minimum.

hugoyvrn replied:
Where can I found the openCV minmial version? I do not found it on the documentation but I think `find_package(OpenCV 4.2.0 REQUIRED)` will be good.

JaouadROS replied:
[ROS2 humble](https://www.ros.org/reps/rep-2000.html#humble-hawksbill-may-2022-may-2027) is shipped with `OpenCV 4.5.4`, so you can have that version as your minimum but if it works with `4.2.0` then it should be good.

JaouadROS reviewed (changes requested):
Here is my first review.

JaouadROS commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -0,0 +1,138 @@
+  const float ThetaShift = 1.5 * 180;
+  const cv::Point origin(image_width / 2, ranges.size());
+
+  for (int r = 0; r < ranges.size(); r++) {  // TODO(??, optimize for cuda)
+    std::vector<cv::Point> pts;
+    cv::ellipse2Poly(origin, cv::Size(r, r), ThetaShift, -bearing, bearing, 1, pts);
+
+    std::vector<cv::Point> arc_points;
+    arc_points.push_back(pts[0]);
+
+    for (size_t k = 0; k < (pts.size() - 1); k++) {  // TODO(??, optimize for cuda)
+      cv::LineIterator it(mono_img, pts[k], pts[k + 1], 4);
+      for (int i = 1; i < it.count; i++, ++it) arc_points.push_back(it.pos());  // TODO(??, optimize for cuda)
+    }
+
+    cv::Mat data_rows_resized;
+    cv::resize(rawDataMat.row(r), data_rows_resized, cv::Size(arc_points.size(), arc_points.size()));
+
+    for (size_t k = 0; k < arc_points.size(); k++)  // TODO(??, optimize for cuda)
+      mono_img.at<dataType>(arc_points[k]) = data_rows_resized.at<dataType>(1, k);
+  }
```
I profiled the code and as expected, this loop is the most consuming. Without adding extra dependency, you can parallelize the code on CPU. Use `cv::parallel_for_`:

```
  cv::parallel_for_(cv::Range(0, ranges.size()), [&](const cv::Range& range) {
    for (int r = range.start; r < range.end; r++) {
      //double r = ranges[i];
      std::vector<cv::Point> pts;
      cv::ellipse2Poly(origin, cv::Size(r, r), ThetaShift, -bearing, bearing, 1, pts);
      
			std::vector<cv::Point> arc_points;
			arc_points.push_back(pts[0]);

			for (size_t k = 0; k < (pts.size() - 1); k++) {  // TODO(??, optimize for cuda)
				cv::LineIterator it(mono_img, pts[k], pts[k + 1], 4);
				for (int i = 1; i < it.count; i++, ++it) arc_points.push_back(it.pos());  // TODO(??, optimize for cuda)
			}

			cv::Mat data_rows_resized;
			cv::resize(rawDataMat.row(r), data_rows_resized, cv::Size(arc_points.size(), arc_points.size()));

			for (size_t k = 0; k < arc_points.size(); k++)  // TODO(??, optimize for cuda)
				mono_img.at<dataType>(arc_points[k]) = data_rows_resized.at<dataType>(1, k);            
    }
    });
```

hugoyvrn committed:
clang tidy cleaning

hugoyvrn committed:
The Magic Number Hunt

hugoyvrn committed:
Refactoring parameter initialisation

hugoyvrn committed:
clang tidy cleaning

hugoyvrn committed:
Adding Jaouad cv::parallel_for_ optimization

hugoyvrn committed:
Adding opencv 4.2.0 requirement

hugoyvrn committed:
Some code cleaning

hugoyvrn committed:
Removing submodule

hugoyvrn commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,2 @@
+Any contribution that you make to this repository
+will be under the Forssea Robotics proprietary license.
```
I do not think this is coherent with BSD3 license. @AugusteBourgois

AugusteBourgois replied:
Right. Replace all occurences of forssea proprietary license with the BSD3 header.

hugoyvrn committed:
Some code cleaning

hugoyvrn committed:
Some code cleaning

AugusteBourgois commented on README.md:
````diff
@@ -42,54 +41,86 @@ cd .. && colcon build
+cd oculus_driver
+git switch master # for development
 ```
+handle the git branch as you want. Please keep in mind you have to pull the reposetory yourself to keep up to date.
````
`repository`

hugoyvrn committed:
The magic numbers hunt continues

hugoyvrn committed:
Code cleaning

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/oculus_sonar_node.hpp:
```diff
@@ -27,7 +27,7 @@
 #include <sensor_msgs/msg/temperature.hpp>
 
-typedef struct {
+struct sonarParameters {
```
Should be `SonarParameters`

AugusteBourgois commented on oculus_ros2/include/oculus_ros2/sonar_viewer.hpp:
```diff
@@ -68,70 +70,65 @@ std::vector<double> linspace(T start_in, T end_in, int num_in) {
-  //     mono_img.at<dataType>(j, i) = 255;
-  //   }
-  // }
+  const int cv_encoding = std::is_same<DataType, uint8_t>::value ? CV_8UC1 : CV_16UC1;
```
This is the same as previously but shorter : the check will be performed each time you execute the function.
What you can do is
```cpp
if constexpr (std::is_same<DataType, uint8_t>::value) {
  mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_8UC1) *
             std::numeric_limits<DataType>::max();
} else {
  mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_16UC1) *
             std::numeric_limits<DataType>::max();
}
```

The compiler will check at build time the condition and only implement the correct version.

AugusteBourgois commented on oculus_ros2/src/sonar_viewer.cpp:
```diff
@@ -15,54 +15,54 @@ SonarViewer::SonarViewer(rclcpp::Node* node) : node_(node) {
 SonarViewer::~SonarViewer() {}
 
-// void SonarViewer::stream_and_filter(const int &width,
+// void SonarViewer::streamAndFilter(const int &width,
```
I don't know if we should leave the filtering code inside this code. I think the filtering should occur somewhere else.

hugoyvrn replied:
In an other C++ class in oculus_ros2 package or in an other repo ?

hugoyvrn committed:
Code cleaning

hugoyvrn commented on oculus_ros2/include/oculus_ros2/oculus_viewer_node.hpp:
```diff
@@ -0,0 +1,32 @@
+
+#ifndef OCULUS_ROS2__OCULUS_VIEWER_NODE_HPP_
+#define OCULUS_ROS2__OCULUS_VIEWER_NODE_HPP_
+
```
@AugusteBourgois I still have `src/oculus_sonar/oculus_ros2/include/oculus_ros2/oculus_viewer_node.hpp:41:  Found C system header after other header. Should be: oculus_viewer_node.h, c system, c++ system, other.  [build/include_order] [4]` when I am runing pre commit

hugoyvrn committed:
Code cleaning

hugoyvrn committed:
Code cleaning

hugoyvrn committed:
Fix communication with flag

hugoyvrn committed:
Various fixies and upgrades

hugoyvrn committed:
FSM for run and overheating fix

hugoyvrn committed:
Removing data_depth

hugoyvrn committed:
Frequency default value fix

hugoyvrn committed:
Removing gain_assist in high frequency

hugoyvrn committed:
Code cleaning

hugoyvrn commented:
I have updated the code, mainly to fix a few bugs. I've removed some non-working parameters (`data_depth` and `gain_assist` in high frequency), these bugs will be solved in a future PR.
